Add compositional recommendations to the CopyButton docs

jorytindall opened the pull request:
### :pushpin: Summary

If merged, this PR adds compositional recommendations for how to use the CopyButton with form inputs. Some notes that I'm looking for feedback on:

- Do these recommendations make sense?
- I attempted to illustrate some of the examples in code (with inline CSS 🤢), is this the correct strategy for this?

I intentionally kept this documentation with the CopyButton rather than cross linking between the form patterns because it seemed more scoped to the CopyButton. There's potentially a way to cross link between this, the form patterns, and the input documentation, but that might be a bit of complex web.

**Preview:** https://hds-website-git-jt-copy-button-composition-docs-hashicorp.vercel.app/components/copy/button

### :link: External links

<!-- Issues, RFC, etc. -->
Jira ticket: [HDS-2315](https://hashicorp.atlassian.net/browse/HDS-2315)

***

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.


[HDS-2315]: https://hashicorp.atlassian.net/browse/HDS-2315?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

KristinLBradley commented on website/docs/components/copy/button/partials/guidelines/guidelines.md:
```diff
@@ -35,3 +35,42 @@ Placing a Copy Button close to the content explicitly communicates what is being
+
+![Example spacing of a control and CopyButton](/assets/components/copy/copy-button-composition-input-spacing.png)
+
+This ensures that other form primitives (helper text, error, etc) will span the combined with of the input control and the CopyButton.
```
```suggestion
This ensures that other form primitives (helper text, error, etc) will span the combined width of the input control and the CopyButton.
```

KristinLBradley commented on website/docs/components/copy/button/partials/code/composition.md:
```diff
@@ -0,0 +1,49 @@
+  <F.Legend>Cluster name</F.Legend>
+  <F.HelperText>This is some example helper text.</F.HelperText>
+  <F.Control>
+    <div style="display: flex; flex-direction: row; gap: 8px; align-items: flex-start;">
```
I would not add inline styles like this as I don't think it's something we would want to model and encourage in examples which consumers are likely to copy and paste.

You can instead add custom styles for use within the Web docs. For an example, see the [the SideNav](https://helios.hashicorp.design/components/side-nav?tab=code#layout) and custom styles for it in this file: "website/app/styles/pages/components/side-nav.scss"

jorytindall replied:
That's what I thought might be a better solution, thanks for the feedback! Fixed in https://github.com/hashicorp/design-system/pull/1876/commits/625a30d4f57ba08d675740c412433b641560fb27

KristinLBradley reviewed (commented):
Looking good! Added some minor feedback.

heatherlarsen commented on website/docs/components/copy/button/partials/code/composition.md:
```diff
@@ -0,0 +1,49 @@
+## Composition with form inputs
```
Creating another section under ComponentAPI isn't a pattern we use elsewhere on the docs website. It also feels relevant enough to be listed as a child of the "How to use this component" section. What do you think about nesting it there instead?

didoo replied:
+1: this belongs to the "how to use" section

jorytindall replied:
I initially had it under "How to use", but moved it to a new section because it seemed to be leaning a bit more towards a separate pattern (not being used in isolation). I've moved this back under "How to use" now.

heatherlarsen reviewed (commented):
Looking good! I do think it's worth adding a link somewhere on the Form patterns page back to this information though.

didoo commented on website/app/styles/pages/components/copy-button.scss:
```diff
@@ -0,0 +1,16 @@
+// COMPONENTS > COPY/COPY-BUTTON
+
+#show-content-components-copy-button {
+  .doc-copy-button-composition-input-control {
```
👍

didoo commented on website/docs/components/copy/button/partials/guidelines/guidelines.md:
```diff
@@ -35,3 +35,42 @@ Placing a Copy Button close to the content explicitly communicates what is being
+
+This ensures that other form primitives (helper text, error, etc) will span the combined width of the input control and the CopyButton.
+
+!!! Info
+
+Visit the [code](/components/copy/button?tab=code#composition-with-form-inputs) tab for contextual examples of this type of composition.
+!!!
+
```
Is this really needed? It's the same for all the other components.

```suggestion
```

jorytindall replied:
Removed in https://github.com/hashicorp/design-system/pull/1876/commits/032c889aa8683a810599444fb7ecb63f452edc2c

didoo reviewed (commented):
Left a few comments/suggestions

heatherlarsen commented on website/docs/components/copy/button/partials/guidelines/guidelines.md:
```diff
@@ -35,3 +35,42 @@ Placing a Copy Button close to the content explicitly communicates what is being
+
+This ensures that other form primitives (helper text, error, etc) will span the combined width of the input control and the CopyButton.
+
+!!! Info
+
+Visit the [Form patterns](/patterns/form-patterns) documentation for more details around combining multiple form elements consistently.
+!!!
```
This doesn't feel like it's providing value related to the original ticket. What I was thinking is that there would be a new section or note added to the Form pattern page that linked back to the new CopyButton documentation. Sorry if my initial comment was unclear.

jorytindall commented:
@heatherlarsen I've done a couple of things to follow up on this:

- I removed that specific link/reference to the pattern docs because it was a bit generic, easy to add back in if necessary
- I added a section in the Form patterns guidelines for Composition with other components, and subsequently a section for copying the input value. My thought process behind this is:
    - Creating a section where it probably makes sense to add new content in the future. This can be separated out if necessary into it's own file, potentially it's own tab if necessary
    - Labelling the section by user need, as opposed to the specific combination of components (e.g., Input control with CopyButton). I did this specifically because it's more focused on the problem/end-user outcome, vs the specific combination of components.

This results in a bit of new content, and probably opens up some other questions about how to organize this type of compositional content (kind of in a gray area about being a separate pattern vs being tacked on to something else).

WDYT?

heatherlarsen commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -247,3 +247,15 @@ For complex forms, indicate **required** fields. This is the most explicit and t
 
 Don't mix required and optional labels, stick to one or the other. Using one method can imply that fields without a label are the inverse of whatever method you choose, though expecting the user to carry this knowledge through a complex form can cause unnecessary [validation](?tab=validation) errors when required fields aren't labeled.
+
+## Composition with other components
+
+While form patterns should generally be comprised of form-specific components, sometimes it may be necessary to compose categorically different components together to satisfy a user need.
```
I'm not sure this content and heading are needed, given that pattern documentation is always about composing various components to make a larger feature.

heatherlarsen commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -247,3 +247,15 @@ For complex forms, indicate **required** fields. This is the most explicit and t
+
+While form patterns should generally be comprised of form-specific components, sometimes it may be necessary to compose categorically different components together to satisfy a user need.
+
+### Copying the input value
+
+If a user needs to copy the value within an input, consider pairing a [CopyButton](/components/copy/button) with the input control. This can reduce the manual friction of copying a value to a single targeted action.
+
+![CopyButton paired with an input control](/assets/components/copy/copy-button-composition-single-line-input-simple.png)
+
+Visit the [CopyButton](/components/copy/button#composition-with-other-components) for more details on this specific composition.
```
In which case, you could consider bumping this whole section up a level.

heatherlarsen replied:
```suggestion
Visit the [CopyButton guidelines](/components/copy/button#composition-with-other-components) for more details on this specific composition.
```

heatherlarsen commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -247,3 +247,15 @@ For complex forms, indicate **required** fields. This is the most explicit and t
+
+While form patterns should generally be comprised of form-specific components, sometimes it may be necessary to compose categorically different components together to satisfy a user need.
+
+### Copying the input value
```
```suggestion
### Copying an input value
```

heatherlarsen reviewed (approved):
I left a few more suggestions on the new content, but it's looking good!

didoo reviewed (approved):
👍
